Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h

```diff
@@ -0,0 +1,438 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto
+ *  LICENSE:     See LICENSE in the top level directory
+ *
+ *  Multi Theft Auto is available from http://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+#pragma once
+
+#include <optional>
+#include <variant>
+#include <SharedUtil.Template.h>
+#include "lua/CLuaOverloadParser.h"
+#include "lua/CLuaFunctionParseHelpers.h"
+#include "lua/CLuaStackChecker.h"
+#include "lua/LuaBasic.h"
+
+struct CLuaFunctionParserBase
+{
+    std::size_t iIndex = 1;
+    std::string strError = "";
+    std::string strErrorFoundType = "";
+
+    template <typename T>
+    inline void typeToNameVariant(SString& accumulator)
+    {
+        using param = typename is_variant<T>::param1_t;
+        if (accumulator.length() == 0)
+            accumulator = typeToName<param>();
+        else
+            accumulator += "/" + typeToName<param>();
+
+        if constexpr (is_variant<T>::count != 1)
+            return typeToNameVariant<is_variant<T>::rest_t>(accumulator);
+    }
+
+    template <typename T>
+    inline SString typeToName()
+    {
+        if constexpr (std::is_same_v<T, std::string>)
+            return "string";
+        else if constexpr (std::is_same_v<T, int> || std::is_same_v<T, float> || std::is_same_v<T, double> || std::is_same_v<T, short> ||
+                           std::is_same_v<T, unsigned int> || std::is_same_v<T, unsigned short>)
+            return "number";
+        else if constexpr (std::is_same_v<T, bool>)
+            return "boolean";
+        else if constexpr (std::is_enum_v<T>)
+            return "enum";
+        else if constexpr (is_specialization<T, std::optional>::value)
+        {
+            using param_t = typename is_specialization<T, std::optional>::param_t;
+            return typeToName<param_t>();
+        }
+        else if constexpr (is_2specialization<T, std::vector>::value)
+            return "table";
+        else if constexpr (is_5specialization<T, std::unordered_map>::value)
+            return "table";
+        else if constexpr (std::is_same_v<T, CLuaFunctionRef>)
+            return "function";
+        else if constexpr (std::is_same_v<T, lua_State*>)
+            return "";            // not reachable
+        else if constexpr (is_variant<T>::value)
+        {
+            SString strTypes;
+            typeToNameVariant<T>(strTypes);
+            return strTypes;
+        }
+        else if constexpr (std::is_pointer_v<T> && std::is_class_v<std::remove_pointer_t<T>>)
+            return GetClassTypeName((T)0);
+        else if constexpr (std::is_same_v<T, dummy_type>)
+            return "";
+    }
+
+    static SString ResolveParameter(lua_State* L, std::size_t index)
```

Comment:
Could also do with a brief explanation. Before reading the source, I interpreted `Resolve` to mean the same thing as `TypeMatch`.

```suggestion
    // ResolveParameter translates an argument (by index) to its typed string representation
    // This is used in messages like `expected string at argument 5, got number (5)`
    //                                                                   ^^^^^^^^^^
    static SString ResolveParameter(lua_State* L, std::size_t index)
```